python/nx-cugraph/nx_cugraph/algorithms/core.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import cupy as cp
+import networkx as nx
+import numpy as np
+import pylibcugraph as plc
+
+import nx_cugraph as nxcg
+from nx_cugraph.utils import networkx_algorithm, not_implemented_for
+
+__all__ = ["k_truss"]
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+@networkx_algorithm
+def k_truss(G, k):
+    if is_nx := isinstance(G, nx.Graph):
+        G = nxcg.from_networkx(G, preserve_all_attrs=True)
+    if nxcg.number_of_selfloops(G) > 0:
+        raise nx.NetworkXError(
+            "Input graph has self loops which is not permitted; "
+            "Consider using G.remove_edges_from(nx.selfloop_edges(G))."
+        )
+    # TODO: create renumbering helper function(s)
+    if k < 3:
```

Comment:
Can you add a comment as an overview of what's being done differently here vs. the else block and why? (eg. PLC isn't even being called, etc.)

Reply:
I thought I did leave a comment: we drop nodes with zero degree. No need to call into pylibcugraph just for that. What comment would you find helpful here?

Reply:
sorry I meant a comment describing the overall goal of the entire `k < 3` block vs. the `else` block. I admittedly don't know much about k_truss (maybe that's the problem?), but I was hoping to get at least a rough idea why they're so different.  For instance, I'm curious why `plc.k_truss_subgraph()` couldn't just be used unconditionally even when k < 3, or why we wouldn't remove zero degree nodes in both cases. However, if this would require an explanation of the k_truss algo itself, then feel free to ignore this.

Reply:
Gotcha. Take a quick look at k_truss documentation:
https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.core.k_truss.html

See: _"[...] where every edge is incident to at least `k-2` triangles."_. So, `k < 3` is an edge case.

> I'm curious why plc.k_truss_subgraph() couldn't just be used unconditionally even when k < 3

It can--we can delete this entire branch if we wanted. PLC behaves the same as networkx for `0 <= k < 3`. We do the branch b/c it is faster, and it lets us check the "stop early" condition if there are no nodes of degree zero. I like fast, and I don't think this branch is excessively complicated.

> why we wouldn't remove zero degree nodes in both cases

k_truss always removes zero-degree nodes. PLC's k_truss does this for us when we call it.

I suppose I can add a couple comments to the code like I just explained here ;) --thanks for the discussion.

Reply:
More detailed code comment added.

Reply:
Oh, I remember the primary reason I made this code branch in the first place and why I think it should exist for now: I was exploring renumbering and wanted to try multiple approaches. I'm sure we'll be doing more renumbering soon.